core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1395,16 +1393,24 @@ private[spark] class DAGScheduler(
   }
 
   private def getAndSetShufflePushMergerLocations(stage: ShuffleMapStage): Seq[BlockManagerId] = {
-    val mergerLocs = sc.schedulerBackend.getShufflePushMergerLocations(
-      stage.shuffleDep.partitioner.numPartitions, stage.resourceProfileId)
-    if (mergerLocs.nonEmpty) {
-      stage.shuffleDep.setMergerLocs(mergerLocs)
+    stage.shuffleDep.synchronized {
+      val oldMergeLocs = stage.shuffleDep.getMergerLocs
+      if (oldMergeLocs.isEmpty) {
+        val mergerLocs = sc.schedulerBackend.getShufflePushMergerLocations(
+          stage.shuffleDep.partitioner.numPartitions, stage.resourceProfileId)
+        if (mergerLocs.nonEmpty) {
+          stage.shuffleDep.setMergerLocs(mergerLocs)
+          mapOutputTracker.registerShufflePushMergerLocations(
+            stage.shuffleDep.shuffleId, mergerLocs)
+        }
+        logDebug(s"Shuffle merge locations for shuffle ${stage.shuffleDep.shuffleId} with" +
+          s" shuffle merge ${stage.shuffleDep.shuffleMergeId} is" +
+          s" ${stage.shuffleDep.getMergerLocs.map(_.host).mkString(", ")}")
+        mergerLocs
+      } else {
+        oldMergeLocs
+      }
     }
-
-    logDebug(s"Shuffle merge locations for shuffle ${stage.shuffleDep.shuffleId} with" +
-      s" shuffle merge ${stage.shuffleDep.shuffleMergeId} is" +
-      s" ${stage.shuffleDep.getMergerLocs.map(_.host).mkString(", ")}")
-    mergerLocs
   }
```

Comment:
We dont need to synchronize here - it is within context of `DAGScheduler` lock
Given this, the change would be simpler:

```
if (stage.shuffleDep.getMergerLocs.nonEmpty) return
val mergerLocs = sc.schedulerBackend.getShufflePushMergerLocations( ... )
if (mergerLocs.nonEmpty) {
  stage.shuffleDep.setMergerLocs(mergerLocs)
  mapOutputTracker.registerShufflePushMergerLocations( ... )
  logDebug( ...)
}
```


We can also rename the method as `configureShufflePushMergerLocations` - since we no longer need to return the value.

Reply:
done